CIP-DappConnector/CIP-dapp-connector.md

```diff
@@ -0,0 +1,226 @@
+
+### cardano.get_utxos(amount: cbor\<value> = undefined, paginate: Paginate = undefined): Promise\<TransactionUnspentOutput[] | undefined>
+
+Errors: `APIError`, `PaginateError`
+
+If `amount` is `undefined`, this shall return a list of all UTXOs (unspent transaction outputs) controlled by the wallet. If `amount` is not `undefined`, this request shall be limited to just the UTXOs that are required to reach the combined ADA/multiasset value target specified in `amount`, and if this cannot be attained, `undefined` shall be returned. The results can be further paginated by `paginate` if it is not `undefined`.
+
+### cardano.get_balance(): Promise\<cbor\<value>>
+
+Errors: `APIError`
+
+Returns the total balance available of the wallet. This is the same as summing the results of `cardano.get_utxos()`, but it is both useful to dApps and likely already maintained by the implementing wallet in a more efficient manner so it has been included in the API as well.
+
+### cardano.get_used_addresses(paginate: Paginate = undefined): Promise\<\<cbor<address>[]>
+
+Errors: `APIError`
+
+Returns a list of all used (included in some on-chain transaction) addresses controlled by the wallet. The results can be further paginated by `paginate` if it is not `undefined`.
+
+### cardano.get_unused_addresses(): Promise\<\<cbor<address>[]>
+
+Errors: `APIError`
+
+Returns a list of unused addresses controlled by the wallet.
+
+### cardano.get_change_address(): Promise\<\<cbor<address>>
+
+Errors: `APIError`
+
+Returns an address owned by the wallet that should be used as a change address to return leftover assets during transaction creation back to the connected wallet. This can be used as a generic receive address as well.
+
+### cardano.sign_tx(tx: cbor\<transaction_body>, metadata: cbor\<auxiliary_data> = undefined): Promise\<cbor\<transaction>>
+
+Errors: `APIError`, `TxSignError`
+
+Requests that a user sign the supplied transaction body. The wallet should ask the user for permission, and if given, try to sign the supplied body and return a signed transaction. If the wallet could not sign the transaction, `TxSignError` shall be returned with the `ProofGeneration` code. Likewise if the user declined it shall return the `UserDeclined` code.
+
+### cardano.sign_tx_input(tx: cbor\<transaction_body>, index: number): Promise\<cbor\<transaction_witness_set>>
+
+Errors: `APIError`, `TxSignError`
+
+Provides lower-level ability for signing that produces the witnesses for just a single input in a transaction. This exists in case dApps need to construct a transaction to satisfy certain properties and the user might only own some of the inputs. The wallet should ask user permission for signing similar to `cardano.sign_tx_input()` and errors are handled in the same way.
+
+### cardano.sign_data(addr: cbor\<address>, sig_structure: cbor\<Sig_structure>): Promise\<Bytes>
```

Comment:
I would prefer to return the vkey/public key together with the signed bytes structure. It makes it easy for verification.
Maybe like this:
```
data_witness = [
    vkey: vkey
    signed_structure: Bytes
]
```
Not sure if as Json or Cbor.
and the return type is then Promise<data_witness>.

Reply:
That's a good idea. It would be pretty easy to add on the wallet's side and would be useful so I'm all for the change.

However, before I change the spec, there is some ambiguity that I noticed with this endpoint. As we pass in an address, there exists the problem for some address types where there can be 2 keys (spending key and staking key) with no way to know which one it is. This isn't the case for Ergo which this spec was initially based on so it made sense to use an address there, but not here.

We could change it to take in a `cbor<stake_credential>` and require the dApps to know how to decompose that from the addresses. This is possible as that data is embedded as a part of the address binary structure and is accessible via `cardano-serialization-lib` as well. At least at first thought this seems better than either
1. defaulting to one or the other (potentially confusing to dApp devs?)
2. having a toggle (some address types have no staking keys - do you error? or default to spending?)

Or do we only want to allow signing by staking keys? I don't think there's a security issue with payment keys as we are signing a very specific CBOR structure that was designed to not have conflicts with signing a tx hash or anything like that. What happens in that case if someone provides an address with no staking component?

This is the one part of the EIP-0012 Ergo spec that wasn't implemented in Yoroi at all and hadn't been finalized as the CIP-0008 message signing spec hadn't been finalized at the time.

Reply:
Actually I guess I had forgotten that in the CIP-0008 spec we can provide a mapping from address <-> keys so we need the full address type too because (quoted from CIP-0008):

>To be able to effectively verify we need two things:
> 1. P1 - (optional) knowledge of the relation of a public key and a Cardano address
> 2. P2 - Knowledge of which algorithm was used to sign
>
> For P1, the mapping of public keys to addresses has two main cases:
> 1. for Shelley addresses, one payment key maps to many addresses (base, pointer, enterprise)
> 2. for Byron addresses, chaincode and address attributes need to be combined with the public key to generate an address
>
> To resolve this, one SHOULD add the full address to the protected header when using a v2 address. The v2 addresses contain the chaincode + attributes and we can verify the address matches combining it with the verification public key.
>`? address: bstr,`
>For P2, we use the alg header and to specify which public key was used to sign, use the cwt protected header.

and this can't be done entirely on the wallet side as a `stake_credential` could map to multiple addresses, unless we want the wallet to just insert a random/arbitrary one, but that seems like confusing behavior coming from the dApp if they take a `stake_credential` from an `address` then use this endpoint and it inserts a different `address` into the headers.

Maybe we could have it pass in both the `cbor<stake_credential>` and `cbor<address>`? And then add an error code to `DataSignError` if they don't match, or even just re-use `DataSignErrorCode.InvalidFormat` with info about the mismatch.

Reply:
I still like the idea of passing in an adress as argument. Yeah right base addresses have 2 credentials. But in my opinion if you want to prove ownership over a base address you care actually only about the payment credential part. So I would standardize that:
`Base Adress` -> Take `Payment Credential`
And if you want to sign with the `Stake Credential` you pass in the `Reward Address`

Reply:
So I'm trying to understand CIP-0008 better and I think we should change the `sign_data` function to:
`sign_data(addr: cbor<address>, payload<hex>) : Promise<Bytes>`
The whole sig struc with the protected/unprotected headers I would build on the wallet side and just pass the payload there and then return the `CoseSign1` or `CoseSign` object from the wallet, which includes the `public key`, `algorithm id` and the `address` in the protected headers. The payload is an utf8 encoded hex string.

Reply:
If we just have the payload, how would the wallet know how to construct the rest of the sig structure? Are we assuming only the CoseSign1 case? What if the dApp wants to put additional fields into the header besides alg/key? I guess one worry is to construct it you might need the pubkey but unless that other endpoint for getting verification keys is implemented (or an alternative) it might be hard to know that as the address only contains the hashes of them.

Reply:
Probably more suitable for the CoseSign1 case right. And yes additional fields would not work, but I think in most cases you just want to make simple data signatures in order to prove ownership or login into a service. (maybe have two endpoints, one for easy signing and one for advanced things? Or rather an external library, that handles the complicated logic and exposes a simple sign function (like web3.js))

It's actually easy for the wallet to identify the public key belonging to an address/key hash. (in the `signTx` function you have to do similar things).
Even if there is a function in the API to get a public key, you still need to dermine first which one you need, and all information you have are addresses from a dApp side, so the wallet needs to know anyway which address/key hash maps to what public key.

Reply:
I think it makes more sense to be more general and accept any sig structure from CIP-0008 and like you said we can use external libraries to simplify the construction. The [message-signing](https://github.com/Emurgo/message-signing) library for example could have simplified code-paths for constructing things in the most common cases which dApps could then use to pass to the connector. Due to other higher priorities we didn't end up using that library ourselves so there's likely some functionality lacking in it that could help.

>It's actually easy for the wallet to identify the public key belonging to an address/key hash. (in the signTx function you have to do similar things).
Even if there is a function in the API to get a public key, you still need to dermine first which one you need, and all information you have are addresses from a dApp side, so the wallet needs to know anyway which address/key hash maps to what public key.

In the 2nd CIP review meeting this Tuesday it was brought up that we either need to change the API from addresses to pubkeys (although IMO this would make it clunkier to use to convert all the keys to addresses for most uses) or provide an API endpoint to get the pubkeys. I proposed one for Address -> PubKey (for addresses the wallet owns only so it could return undefined if it doesn't know it). This could be used for this. @KtorZ also proposed the following:

>Provide a way to retrieve verification keys at a given derivation path. For example getVerificationKey(path : string) : Promise<VerificationKey>. The path could be either a list of levels or a slash-separated string (e.g. "1852'/1815'/0'/0/0)`. That is probably the most flexible option.

in another comment below in this PR.

Reply:
> I proposed one for Address -> PubKey (for addresses the wallet owns only so it could return undefined if it doesn't know it)

Yes I am for this idea. Keep the address enpoint and have this public key endpoint extra.

>  For example getVerificationKey(path : string) : Promise. The path could be either a list of levels or a slash-separated string (e.g. "1852'/1815'/0'/0/0)`

This may be a little bit too flexible, because this would mean the dApp could derive keys/addresses the wallet is not even familiar with or there is a higher chance that the dApp could make something wrong.